Std/Data/List/Perm.lean

```diff
@@ -0,0 +1,889 @@
+    h₁ ++ a :: t₁ ~ h₂ ++ a :: t₂ :=
+  p₁.append (p₂.cons a)
+
+@[simp]
+theorem perm_middle {a : α} : ∀ {l₁ l₂ : List α}, l₁ ++ a :: l₂ ~ a :: (l₁ ++ l₂)
+  | [], _ => Perm.refl _
+  | b :: l₁, l₂ => ((@perm_middle a l₁ l₂).cons _).trans (swap a b _)
+
+@[simp]
+theorem perm_append_singleton (a : α) (l : List α) : l ++ [a] ~ a :: l :=
+  perm_middle.trans <| by rw [append_nil]; apply Perm.refl
+
+theorem perm_append_comm : ∀ {l₁ l₂ : List α}, l₁ ++ l₂ ~ l₂ ++ l₁
+  | [], l₂ => by simp
+  | a :: t, l₂ => (perm_append_comm.cons _).trans perm_middle.symm
+
+theorem concat_perm (l : List α) (a : α) : concat l a ~ a :: l := by simp
+
+theorem Perm.length_eq {l₁ l₂ : List α} (p : l₁ ~ l₂) : length l₁ = length l₂ :=
+  p.rec
+    rfl
+    (fun _x l₁ l₂ _p r => by simp [r])
+    (fun _x _y l => by simp)
+    (fun _p₁ _p₂ r₁ r₂ => Eq.trans r₁ r₂)
+
+theorem Perm.eq_nil {l : List α} (p : l ~ []) : l = [] :=
+  eq_nil_of_length_eq_zero p.length_eq
+
+theorem Perm.nil_eq {l : List α} (p : [] ~ l) : [] = l :=
+  p.symm.eq_nil.symm
+
+@[simp]
+theorem perm_nil {l₁ : List α} : l₁ ~ [] ↔ l₁ = [] :=
+  ⟨fun p => p.eq_nil, fun e => e ▸ Perm.refl _⟩
+
+@[simp]
+theorem nil_perm {l₁ : List α} : [] ~ l₁ ↔ l₁ = [] :=
+  perm_comm.trans perm_nil
+
+theorem not_perm_nil_cons (x : α) (l : List α) : ¬[] ~ x :: l
+  | p => by injection p.symm.eq_nil
+
+@[simp]
+theorem reverse_perm : ∀ l : List α, reverse l ~ l
+  | [] => Perm.nil
+  | a :: l => by
+    rw [reverse_cons]
+    exact (perm_append_singleton _ _).trans ((reverse_perm l).cons a)
+
+theorem perm_cons_append_cons {l l₁ l₂ : List α} (a : α) (p : l ~ l₁ ++ l₂) :
+    a :: l ~ l₁ ++ a :: l₂ :=
+  (p.cons a).trans perm_middle.symm
+
+@[simp]
+theorem perm_replicate {a : α} {n : Nat} {l : List α} :
+    l ~ List.replicate n a ↔ l = List.replicate n a :=
+  ⟨fun p => eq_replicate.2
+    ⟨p.length_eq.trans <| length_replicate _ _, fun _b m => eq_of_mem_replicate <| p.subset m⟩,
+    fun h => h ▸ Perm.refl _⟩
+
+@[simp]
+theorem replicate_perm {a : α} {n : Nat} {l : List α} :
+    List.replicate n a ~ l ↔ List.replicate n a = l :=
+  (perm_comm.trans perm_replicate).trans eq_comm
+
+@[simp]
+theorem perm_singleton {a : α} {l : List α} : l ~ [a] ↔ l = [a] :=
+  @perm_replicate α a 1 l
+
+@[simp]
+theorem singleton_perm {a : α} {l : List α} : [a] ~ l ↔ [a] = l :=
+  @replicate_perm α a 1 l
+
+theorem Perm.eq_singleton {a : α} {l : List α} (p : l ~ [a]) : l = [a] :=
+  perm_singleton.1 p
+
+theorem Perm.singleton_eq {a : α} {l : List α} (p : [a] ~ l) : [a] = l :=
+  p.symm.eq_singleton.symm
+
+theorem singleton_perm_singleton {a b : α} : [a] ~ [b] ↔ a = b := by simp
+
+theorem perm_cons_erase [DecidableEq α] {a : α} {l : List α} (h : a ∈ l) : l ~ a :: l.erase a :=
+  let ⟨_l₁, _l₂, _, e₁, e₂⟩ := exists_erase_eq h
+  e₂.symm ▸ e₁.symm ▸ perm_middle
+
+@[elab_as_elim]
+theorem perm_induction_on {P : List α → List α → Prop} {l₁ l₂ : List α} (p : l₁ ~ l₂) (h₁ : P [] [])
```

Comment:
We should restore the induction principle that was deleted [here](https://github.com/leanprover-community/mathlib4/commit/c5a07b2253b534effb902ccf5492d83480baed3a#diff-8cf784dd8c93e9893e31189155685d581e0e8915443d84c23756701010cc68b4L96), this one is not usable with `induction using`.